src/masternodeman.cpp

```diff
@@ -378,35 +366,46 @@ CMasternode *CMasternodeMan::FindRandomNotInVec(std::vector<CTxIn> &vecToExclude
     LogPrintf("CMasternodeMan::FindRandomNotInVec -- %d enabled masternodes, %d masternodes aren't yet exluded\n", nCountEnabled, nCountNotExcluded);
     if(nCountNotExcluded < 1) return NULL;
 
-    std::vector<CMasternode> vMasternodesShuffled = vMasternodes;
-    std::random_shuffle(vMasternodesShuffled.begin(), vMasternodesShuffled.end(), GetRandInt);
+    // fill a vector of pointers
+    std::vector<CMasternode*> vpMasternodesShuffled;
+    BOOST_FOREACH(CMasternode &mn, vMasternodes) {
+        vpMasternodesShuffled.push_back(&mn);
+    }
+
+    // shuffle pointers
+    std::random_shuffle(vpMasternodesShuffled.begin(), vpMasternodesShuffled.end(), GetRandInt);
     bool fExclude;
 
-    BOOST_FOREACH(CMasternode &mn, vMasternodesShuffled) {
-        if(mn.nProtocolVersion < nProtocolVersion || !mn.IsEnabled()) continue;
+    // loop through
+    BOOST_FOREACH(CMasternode* pmn, vpMasternodesShuffled) {
+        if(pmn->nProtocolVersion < nProtocolVersion || !pmn->IsEnabled()) continue;
         fExclude = false;
         BOOST_FOREACH(CTxIn &txinToExclude, vecToExclude) {
-            if(mn.vin.prevout == txinToExclude.prevout) {
+            if(pmn->vin.prevout == txinToExclude.prevout) {
```

Comment:
This can probably be made much more efficient by putting txinToExclude into a set and using find O(N*M) -> O(N*log(M)) where M = txinToExclude.size() and N = # masternodes.